c/misra/src/rules/DIR-4-6/PlainNumericalTypeUsedOverExplicitTypedef.ql

```diff
@@ -0,0 +1,53 @@
+/**
+ * @id c/misra/plain-numerical-type-used-over-explicit-typedef
+ * @name DIR-4-6: Do not use plain numerical types over typedefs named after their explicit bit layout
+ * @description Using plain numerical types over typedefs with explicit sign and bit counts may lead
+ *              to confusion on how much bits are allocated for a value.
+ * @kind problem
+ * @precision high
+ * @problem.severity error
+ * @tags external/misra/id/dir-4-6
+ *       external/misra/obligation/advisory
+ */
+
+import cpp
+import codingstandards.c.misra
+import codingstandards.cpp.BuiltInNumericTypes
+
+class BuiltInNumericType extends BuiltInType {
+  BuiltInNumericType() {
+    /* Exclude the plain char because it does not count as a numeric type */
+    this.(CharType).isExplicitlySigned()
+    or
+    this.(CharType).isExplicitlyUnsigned()
+    or
+    this instanceof BuiltInIntegerType
+    or
+    this instanceof FloatType
+    or
+    this instanceof DoubleType
+    or
+    this instanceof LongDoubleType
+  }
+}
+
+predicate forbiddenBuiltinNumericUsedInDecl(Variable var, string message) {
+  var.getType() instanceof BuiltInNumericType and
+  not var instanceof ExcludedVariable and
+  message = "The type " + var.getType() + " is not a fixed-width numeric type."
+}
+
+predicate forbiddenTypedef(TypedefType typedef, string message) {
+  typedef.getBaseType() instanceof BuiltInNumericType and
+  not typedef.getName().regexpMatch("u?(int|float)(4|8|16|32|64|128)_t") and
```

Comment:
We can actually extract the bits using the regex and test this against the size of the base type to detect incorrect typedefs.

Reply:
That would be a nice addition, but isn't guaranteeing the correctness of the named size a bit out of scope of this rule?

Reply:
Currently you could write your own typedef in your source file with the wrong type, and the query would consider that "acceptable", even though it clearly violates the rule.

In any case, the rule actually says the following:
> For C99, the types provided by <stdint.h> should be used.
I think if we confirmed that the definitions were from `stdint.h`, we wouldn't need to check the sizes. However, that is somewhat trickier (because the definitions may not be in `stdint.h` directly, just included from those definitions).

For now, I think it's easiest just to check the type sizes as Remco suggested.

Reply:
Hmm, I could not find the quoted part in my copy of the MISRA C:2012 document. Could you let me know if there is an additional source I should look into? 

In the meantime, I added the check.